app/services/notifications/notifiable_action/send.rb

```diff
@@ -25,7 +25,16 @@ def call
         json_data[:organization] = organization_data(notifiable.organization) if notifiable.organization_id
 
         notifications_attributes = []
-        notifiable.followers.sort_by(&:updated_at).last(10_000).reverse_each do |follower|
+
+        # If a user was mentioned in the article, they will have already received a mention.
+        # We explicitly need to exclude them from the article_followers array if they already
+        # have a mention in order to avoid sending a user multiple notifications for one article.
+        article_followers = notifiable.followers.reject do |follower|
+          users_with_mentions_on_article = notifiable.mentions&.pluck(:user_id)
+          users_with_mentions_on_article.include?(follower.id)
+        end
+
+        article_followers.sort_by(&:updated_at).last(10_000).reverse_each do |follower|
```

Comment:
I wonder if we should move the `sort` and `last` into the DB instead (`order(:updated_at: :desc).limit(10_000)`) so `article_followers` already contains the correctly sorted and limited collection. If we add the correct `joins` we could probably replace the `reject` with a `where.not()`.

If that is too complicated the following would still work and at least have the `reject` only run on a maximum of 10k followers. This may mean we're not sending out exactly 10k notifications but I think that's ok:

```ruby
article_followers = notifiable.followers.order(:updated_at: :desc).limit(10_000).reject do |follower|
  # same as above
end
# We wouldn't need reverse_each anymore since the newest would be on top already
article_followers.each do |follower| 
```

Reply:
Tweaked this in my latest commit; I like the idea of moving some of these methods into DB calls directly, that seems like a clear/obvious improvement!

Reply:
Ok, that makes sense, sorry I missed that.